lib/active_record/associated_object/object_association.rb

```diff
@@ -9,15 +11,34 @@ def extend_source_from(chunks, &block)
     end
   }
 
-  def has_object(*names, **callbacks)
-    extend_source_from(names) do |name|
-      "def #{name}; @#{name} ||= #{self.name}::#{name.to_s.classify}.new(self); end"
-    end
+  included do
+    class_attribute :associated_object_ivar_names,
+      default: [],
+      instance_accessor: false,
+      instance_predicate: false
+  end
+
+  class_methods do
+    def has_object(*names, **callbacks)
+      self.associated_object_ivar_names += names.map { |n| :"@#{n}" }
+
+      extend_source_from(names) do |name|
+        "def #{name}; @#{name} ||= #{self.name}::#{name.to_s.classify}.new(self); end"
+      end
```

Comment:
Note: I'm also applying the change from #19 to remove the needless `#{self.name}::` here.

```suggestion
      extend_source_from(names) do |name|
        "def #{name}; (@associated_objects ||= {})[:#{name}] ||= #{name.to_s.classify}.new(self); end"
      end
```